### ISSUE TITLE
Feature request: ability to enable Docker Buildkit for `sam build`

### ISSUE THREAD

**Issue by ghost** [aws/aws-sam-cli#3972]
### Describe your idea/feature/enhancement

I wish SAM CLI would enable me to enable Docker Buildkit when running `sam build` for a function with `PackageType: Image`.

This would enable me to use BuildKit features such as cache mounts for faster and more efficient builds.

### Proposal

A very specific approach would be to add a new option along the lines of `--use-docker-buildkit` or something like that.

That said, something more general might be a good idea.

For example, I tried to enable buildkit by running `export DOCKER_BUILDKIT=1 && sam build` but it didn’t work.

IIRC `docker build` supports various other environment variables, so it might be a good idea to have a general purpose mechanism to set environment variables that’d be exposed to the `docker build` process.

One way to do this might be to simply forward/inherit the environment vars from the user’s shell to/into the `docker build` process.

Another way could be to add a CLI option like `--docker-build-env DOCKER_BUILDKIT=1,FOO=bar`

### Additional Details

This _will not_ require any updates to the [SAM Spec](https://github.com/awslabs/serverless-application-model).

**Comment by aviflat** [comment]
Huh, I just found docker-py#2230 and it looks like bad news… 😞

**Comment by qingchm** [comment]
I know it's going to be an interesting feature to support but since our builds are based on docker API for python, unless docker supports buildkit on python api I don't see a quick solution to this!

**Comment by aviflat** [comment]
Makes sense. Thanks.

Maybe it’d help if maintainers of downstream libs would comment on that issue? Or otherwise somehow convey to the maintainers of `docker-py` that support for Buildkit would be great?

Thanks!

**Comment by qingchm** [comment]
@aviflat Yep I would assume opening an issue against their repository would be the way to go!

**Comment by aviflat** [comment]
Well, since docker/docker-py#2230 already exists, I was suggesting that one of the maintainers of this project post a comment to that issue, voicing support for the feature.

**Comment by mildaniel** [comment]
Since there isn't much we can do on our side, I will close this for now but please re-open when this is actionable for us.

**Comment by remidebette** [comment]
Hi,
Alternatively, for using the docker cli without reimplementing the wheel, the project https://github.com/gabrieldemarmiesse/python-on-whales readily provides a python wrapper on the CLI.

It works fine with buildx

**Comment by antonio-carapcea-softwire** [comment]
Any news about this?

**Comment by bnusunny** [comment]
We are considering this request. The initial idea is to allow users to opt-in for using docker cli to build images.

**Comment by royassis** [comment]
@bnusunny any news?

**Comment by faraujo427** [comment]
I'm thinking about working a PR for this feature. I plan to use https://github.com/gabrieldemarmiesse/python-on-whales, as mentioned.

My first thought is to add a dictionary property `DockerBuildx` on `Metadata` and support the following properties:

```yaml
DockerBuildx:
  Tool: str # One of "docker", "podman", or "nerdctl"
  Builder: str
  Secrets: str
  Labels: dict
```

Any thoughts and sugestions are welcome

**Comment by royassis** [comment]
Python on whales is great.
I am using it as well.

**Comment by reedham-aws** [comment]
> I'm thinking about working a PR for this feature. I plan to use https://github.com/gabrieldemarmiesse/python-on-whales, as mentioned.

I think python-on-whales could work, that's basically what we were planning on doing except our plan is to call the CLI directly as a subprocess. We wouldn't want to replace all of docker-py with python-on-whales, but we'd have a separate flow using the CLI in some way. I think the main concern would be relying on that project as a dependency when it's not as commonly used as the Docker CLI or docker-py.

> My first thought is to add a dictionary property `DockerBuildx` on `Metadata` and support the following properties

I think this could work as well, although currently our plan is to use a build flag that would be something akin to `sam build --with-buildkit`.

This feature is something that is in active development by our team right now and has been having design reviews internally, so if you want to hold off @francilioaraujo, we'll have a PR out soon. I don't want to over-promise, but definitely by the end of February we are looking at releasing this feature. I will update this thread as we open the PR, or to respond to any questions.

**Comment by reedham-aws** [comment]
Another thing worth noting is that we also support the Finch container engine, so python-on-whales wouldn't work for users that only have `finch` CLI. This wasn't a problem before as `docker-py` works agnostically with the container engine, whereas CLI usage relies on the executable existing.

**Comment by reedham-aws** [comment]
Hello, we have merged this feature into the develop branch and it will be fully released in SAM CLI for release 1.156. This will be sometime next week, ~2026-03-18.
